crates/extensions/c8y_firmware_manager/src/actor.rs

```diff
@@ -0,0 +1,723 @@
+        let child_id = timeout.event.child_id;
+        let operation_id = timeout.event.operation_id;
+
+        if let Some(_operation_state) = self
+            .active_child_ops
+            .get(&OperationKey::new(&child_id, &operation_id))
+        {
+            self.fail_operation_in_cloud(
+                &child_id,
+                Some(&operation_id),
+                &format!("Child device {child_id} did not respond within the timeout interval of {}sec. Operation ID={operation_id}", self.config.timeout_sec.as_secs()),
+                message_box,
+            ).await
+        } else {
+            // Ignore the timeout as the operation has already completed.
+            Ok(())
+        }
+    }
+
+    // This function can be removed once we start using operation ID from c8y.
+    async fn validate_same_request_in_progress(
+        &mut self,
+        smartrest_request: SmartRestFirmwareRequest,
+    ) -> Result<(), FirmwareManagementError> {
+        let firmware_dir_path = self.config.validate_and_get_firmware_dir_path()?;
+
+        for entry in fs::read_dir(firmware_dir_path.clone())? {
+            match entry {
+                Ok(file_path) => match FirmwareOperationEntry::read_from_file(&file_path.path()) {
+                    Ok(recorded_entry) => {
+                        if recorded_entry.child_id == smartrest_request.device
+                            && recorded_entry.name == smartrest_request.name
+                            && recorded_entry.version == smartrest_request.version
+                            && recorded_entry.server_url == smartrest_request.url
+                        {
+                            return Err(FirmwareManagementError::RequestAlreadyAddressed);
+                        }
+                    }
+                    Err(err) => {
+                        warn!("Error: {err} while reading the contents of persistent store directory {}",
+                            firmware_dir_path.display());
+                        continue;
+                    }
+                },
+                Err(err) => {
+                    warn!(
+                        "Error: {err} while reading the contents of persistent store directory {}",
+                        firmware_dir_path.display()
+                    );
+                    continue;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    async fn fail_operation_in_cloud(
+        &mut self,
+        child_id: &str,
+        op_id: Option<&str>,
+        failure_reason: &str,
+        message_box: &mut FirmwareManagerMessageBox,
+    ) -> Result<(), FirmwareManagementError> {
+        error!(failure_reason);
+        let op_state = if let Some(operation_id) = op_id {
+            self.remove_status_file(operation_id)?;
+            self.remove_entry_from_active_operations(&OperationKey::new(child_id, operation_id))
+        } else {
+            ActiveOperationState::Pending
```

Comment:
That logic seems probelematic. An `operation_key` may not be present in the `active_child_ops` map even when the operation has already completed with successful/failed response, upon which the entries are removed from the `active_child_ops` map. In that case, assuming that operation to be `Pending` would be wrong.